### PR TITLE
docs(manager:regex): Improve regex manager documentation

### PR DESCRIPTION
**Pull request by TWiStErRob**
## Changes

While the [page for regex manager](https://docs.renovatebot.com/modules/manager/regex/) gives a list of possible fields, it's very unclear what the values should be for each of those fields.

[**Live rendered preview**](https://github.com/TWiStErRob/renovate/blob/patch-2/lib/modules/manager/regex/readme.md) for this branch (always HEAD).

## Context

I wanted to simply match `const val VERSION_KOTLIN: String = "1.4.32"` kotlin code and use it as a Gradle `org.jetbrains.kotlin:kotlin-stdlib` dependency. I'm hoping that these links would help others to faster navigate the docs. I'm willing to improve further, because even the linked pages only contain some basic descriptions, but not much about the actual potential values of each, or where to find those.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [inline lib/modules/manager/regex/readme.md:24]: I haven't looked into how to do a self-link yet without hardcoding an absolute URL.

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/regex/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/regex/readme.md:16]:
```suggestion
- Which [`datasource`](https://docs.renovatebot.com/modules/datasource/#supported-datasources) to look up (e.g. [npm](https://docs.renovatebot.com/modules/datasource/#npm-datasource), [Docker](https://docs.renovatebot.com/modules/datasource/#docker-datasource), [GitHub tags](https://docs.renovatebot.com/modules/datasource/#github-tags-datasource)
```

Drop the `etc.` entry. There's a general `datasource` link at the beginning of the sentence already. 😉

File: lib/modules/manager/regex/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/regex/readme.md:19]:
```suggestion
<!-- prettier-ignore -->
!!! note
    The capture groups below are similar to [available fields](https://docs.renovatebot.com/templates/#other-available-fields), but these available fields are not available in templates below, because most of those are coming from managers, and `regex` is a manager.
```

Let's make an admonition for this piece of text, so it's highlighted.

File: lib/modules/manager/regex/readme.md
Comment (HonkingGoose) [reply lib/modules/manager/regex/readme.md:24]: We can use _relative_ links, but I always forget how that works. 🙈 

@viceice can you help here?

File: lib/modules/manager/regex/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/regex/readme.md:25]:
```suggestion
- You can optionally have a `packageName` capture group or a [`packageNameTemplate` config field](https://docs.renovatebot.com/configuration-options/#packagenametemplate) if it differs from `depName`
```

Add the URL to the link.

File: lib/modules/manager/regex/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/regex/readme.md:30]:
```suggestion
- You can optionally have an [`extractVersion`](https://docs.renovatebot.com/configuration-options/#extractversion) capture group or an [`extractVersionTemplate` config field](https://docs.renovatebot.com/configuration-options/#extractversiontemplate)
```

Let's add the URL here as well.

File: lib/modules/manager/regex/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/regex/readme.md:33]:
```suggestion
- You can optionally have a [`registryUrl` capture group](https://docs.renovatebot.com/configuration-options/#registryurls) or a [`registryUrlTemplate` config field](https://docs.renovatebot.com/configuration-options/#registryurltemplate), if it's a valid URL, it will be converted to the [`registryUrls` field](https://docs.renovatebot.com/configuration-options/#registryurls) as a single-length array
```

MkDocs or Material for MkDocs doesn't know how render a list like this:

- You can optionally have a registryUrl capture group or a registryUrlTemplate config field
  - If it's a valid URL, it will be converted to the registryUrls field as a single-length array.
  
The list is also broken on the published [`regex` manager page](https://docs.renovatebot.com/modules/manager/regex/#required-fields).

Let's put the `if` part on the same line, so it renders properly.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/modules/manager/regex/readme.md
Comment (viceice) [reply lib/modules/manager/regex/readme.md:24]: should work, sample:
https://github.com/renovatebot/renovate/blob/41fd7fbf46ca9e3a46f5ace2a1de213929d9e347/docs/usage/getting-started/private-packages.md#L344

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md:19]: Is the sentence true? I was just assuming, didn't actually test. I think it's true, because it's a chicken/egg problem.

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md:24]: I'm not sure these relatives are a good idea. 
image

All we get is warnings upon warnings (IntelliJ IDEA):
image

We also lose context, e.g. `../../datasource/` assumes that this part of the code is actually hosted at the url `/modules/manager/regex/` and hides the fact that it's referring to `/modules/datasource`. I came at this from the maintenance perspective. If there's any refactor of docs, it'll be hard to update, or even discover that it's linked.

The linked "GitLab Runner" example links to the same folder `/docs` in source and in prod, but the links I added here pretty much link to virtual folders in prod (i.e. configuration-options).

How about going with absolute (i.e. host-relative) links?
```suggestion
- You must have either a `depName` capture group or a [`depNameTemplate` config field](/configuration-options/#depnametemplate)
```

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md]: https://github.com/renovatebot/renovate/pull/15067/commits/7012fda26706ad33afea5060ca28c4d8e288a4a8

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md]: https://github.com/renovatebot/renovate/pull/15067/commits/3712aba65d99f523680b98004e78db6787633a76

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md]: https://github.com/renovatebot/renovate/pull/15067/commits/943af79941ad24eabbd77d5f74c5eda33a3be57e

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md]: https://github.com/renovatebot/renovate/pull/15067/commits/943af79941ad24eabbd77d5f74c5eda33a3be57e

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md]: https://github.com/renovatebot/renovate/pull/15067/commits/29200839ba01c0e61d8481436b27d4ac04d084c2

Comment (viceice) [reply lib/modules/manager/regex/readme.md]: please read https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#resolve-review-comments-instead-of-commenting to reduce noise

Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md]: 🤔 you still get an email about resolution, and also how do you take a look if I had done the right thing? (There's a suggestion, but I might've not done it exactly like that. For example in this particular comment I didn't join two sentences with a comma, I kept the sentences instead, because otherwise it becomes a run-on sentence.) Anyway, I'll press the button without linking the relevant changes in the future if that helps.

Comment (viceice) [reply lib/modules/manager/regex/readme.md]: We review the new changes and do new comments if required. We never review single commits. So `done` or `<git-sha>` comments  are useless for us.

Of cause you should comment, if something needs further discussion. So don't resolve without any code change or comment, 😉 

Also github has a nice option to batch suggestions to a single commit on the review page  😉

Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md]: Ah, ok, good faith and end result matters principles 😅. Thanks for the clarification!

File: lib/modules/manager/regex/readme.md
Comment (HonkingGoose) [reply lib/modules/manager/regex/readme.md:26]: Whatever we do, it should be easy and simple to maintain, and we shouldn't end up with an IDE filled with warnings.

I don't think we can use _absolute_ links though. Quote from the [manual for MkDocs on linking to pages](https://www.mkdocs.org/user-guide/writing-your-docs/#linking-to-pages):

> Warning
> 
> Using absolute paths with links is not officially supported. Relative paths are adjusted by MkDocs to ensure they are always relative to the page. Absolute paths are not modified at all. This means that your links using absolute paths might work fine in your local environment but they might break once you deploy them to your production server.

Comment (viceice) [reply lib/modules/manager/regex/readme.md:26]: the problem is probably that we shift the modules pages around before building the final docs.

so i would suggest that we use repo relative links. and when copying files we fix the links to new relationship.

**Comment by MaronHatoum** [comment]
hi @TWiStErRob 
Are any more updates needed here? or are you waiting for maintainers to review?

**Comment by chludwig-haufe** [comment]
Proposal from discussion #17604:

> ### depName vs. packageName
> By default, the fields `depName` and `packageName` have the same value. You may override the value of `packageName` by means of either a `packageName` capture group or a `packageNameTemplate`.
>
> `packageName` is used to look up a dependency in the respective datasource. For instance, you use the package name `python/cpython` in order to refer to the CPython releases in the `github-tags` datasource or `my.private.registry/repo/image-name` to refer to an image `repo/image-name` in the registry `my.private.registry` using the `docker` datasource.
>
> `depName`, on the other hand, is the name assigned to a dependency. It is used, e.g., to create the [default branch topic](https://docs.renovatebot.com/configuration-options/#branchtopic). Despite their name, the package rule config options [`matchPackageNames`](https://docs.renovatebot.com/configuration-options/#matchpackagenames), [`matchPackagePatterns`](https://docs.renovatebot.com/configuration-options/#matchpackagepatterns), and [`matchPackagePrefixes`](https://docs.renovatebot.com/configuration-options/#matchpackageprefixes) match against the value of `depName`, not `packageName`.

However, as pointed out by @viceice, there are plans in #16012 to make `packageName` the primary filter and `depName` "just" a visual name.

File: lib/modules/manager/regex/readme.md
Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md:26]: @MaronHatoum I think I would be happy to merge, but we reached a stalemate here, it's really not in scope of this PR to fix how MkDocs, but I don't know what to do, how to link to those other parts of the docs in a way that would make this PR approved.

Comment (TWiStErRob) [reply lib/modules/manager/regex/readme.md:26]: > when copying files we fix the links to new relationship.

@viceice to me this sounds like there needs to be some tooling improvements before we can merge? otherwise we end up with broken links.

**Comment by TWiStErRob** [comment]
@MaronHatoum @chludwig-haufe I think we should merge this before making any further improvements to this page. (Which should be in scope of the PR changing the logic proposed in https://github.com/renovatebot/renovate/issues/16012).

File: lib/modules/manager/regex/readme.md
Comment (viceice) [reply lib/modules/manager/regex/readme.md:26]: use absolute urls for now, as we don't have any tooling yet to fix the relative paths for publishing

File: lib/modules/manager/regex/readme.md
Comment (viceice) [inline lib/modules/manager/regex/readme.md:16]:
```suggestion
- Which [`datasource`](https://docs.renovatebot.com/modules/datasource/#supported-datasources) to look up (e.g. [`npm`](https://docs.renovatebot.com/modules/datasource/#npm-datasource), [`docker`](https://docs.renovatebot.com/modules/datasource/#docker-datasource), [`github-tags`](https://docs.renovatebot.com/modules/datasource/#github-tags-datasource))
```

use the datasource id's, as those must be used in those fields

File: lib/modules/manager/regex/readme.md
Comment (viceice) [inline lib/modules/manager/regex/readme.md:21]: It still valid, as those fieds are not yet available, you can only use captured groups from regex match.

File: lib/modules/manager/regex/readme.md
Comment (viceice) [inline lib/modules/manager/regex/readme.md:26]: Is it clear that you can have both and use the all capture groups in the `...Template` fields here?

**Comment by rarkins** [comment]
Closing due to staleness and conflicts

**Pull request closed without merging**
